### PR TITLE
Command registry refactor

### PR DESCRIPTION
**Pull request by braydencw1**

### CODE REVIEW

File: bot/command_registry.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		commands: make(map[string]*Command),
+	}
+}
+
+func (c *CommandRegistry) Register(name string, command func(c CommandCtx) error, numArgs int) {
+	c.commands[name] = &Command{
+		fn:              command,
+		numRequiredArgs: numArgs,
+	}
+}
```
Comment (b1naryth1ef) [inline bot/command_registry.go:37]: nit: newline after function

File: bot/bot.go [hunk 2]
```diff
@@ -43,8 +43,7 @@ func HandleMessageEvents(discord *discordgo.Session, msg *discordgo.MessageCreat
 		discord.ChannelMessageSend(msg.ChannelID, "https://imgur.com/K7lTDGU")
 	}
 
-	AddGriefer(discord, msg)
-	HandleCommands(discord, msg)
+	addGriefer(discord, msg)
```
Comment (b1naryth1ef) [inline bot/bot.go]: For the future it probably makes sense to register this as a separate event handler (like we did with CommandRegistry.HandleMessage [here](https://github.com/braydencw1/venova/pull/17/files#diff-8a7fed410ae8fe7d4deb7e8a06ef882c5e036ab39af60211c2fa433168f3ff15R50).

File: bot/command_registry.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+func (c *CommandRegistry) HandleMessage(s *discordgo.Session, msg *discordgo.MessageCreate) {
+	if msg.Author.ID != s.State.User.ID {
+		log.Printf(msg.Author.Username + ": " + msg.Content)
```
Comment (b1naryth1ef) [inline bot/command_registry.go]: nit: lets move this somewhere else (`HandleMessageEvents`?)

File: bot/command_registry.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+			return
+		}
+	}
+	ctx := CommandCtx{
+		Session: s,
+		Message: msg,
+		Args:    args,
+	}
+	// Found mcCmd and executing
+	err := command.fn(ctx)
```
Comment (b1naryth1ef) [inline bot/command_registry.go]: q: do we want to `go` run this?

File: bot/command_registry.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		log.Printf("%s", err)
```
Comment (b1naryth1ef) [inline bot/command_registry.go]: q: do we want to tell the user the command failed? or leave that up to each command?

File: bot/dnd.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+			err := db.InsertPlayDate(t, currRoleId)
+			if err != nil {
+				return fmt.Errorf("error inserting into table %w", err)
+			}
+			ctx.Reply("The Date has been updated.")
+			return nil
+		}
+	}
+	return nil
+}
```
Comment (b1naryth1ef) [inline bot/dnd.go:31]: nit: newline after function

File: bot/mc.go [hunk 7]
```diff
@@ -4,43 +4,116 @@ import (
...
 
 	"github.com/gorcon/rcon"
 )
 
-func restartMinecraft() {
+func manageMinecraftCmd(ctx CommandCtx) error {
+	msg := ctx.Message.Message
+	sess := ctx.Session
+	args := ctx.Args
+	actions := map[string]struct {
```
Comment (b1naryth1ef) [inline bot/mc.go]: Make this a var outside the function:


```go
type minecraftAction struct { ... }
var minecraftActions = map[string]minecraftAction{...}
```

File: bot/mc.go [hunk 8]
```diff
@@ -4,43 +4,116 @@ import (
...
+				}
+				time.Sleep(time.Second * 5)
+				sess.ChannelMessageEdit(msg.ChannelID, mcMsg.ID, action.Message)
+			}()
+			return nil
+		}
+	} else {
+		ctx.Reply("Action is not available")
+		return nil
+	}
```
Comment (b1naryth1ef) [inline bot/mc.go]: You should get into the habit of identifying and cleaning up if statements like this. Whenever you see a huge `if` block followed by a very small else that breaks control-flow (return/break/continue) somehow, that should probably be reworked to have the else statement checked before. In this example it's much cleaner (and safer when we touch it in the future) to write this as:

```suggestion
	action, exists := actions[args[0]]
	if !exists {
		ctx.Reply("Action is not available")
		return nil
	}
	if memberHasRole(msg.Member, mcRoleId) || msg.Author.ID == morthisId {
		mcMsg, _ := sess.ChannelMessageSend(msg.ChannelID, "Attempting to modify the minecraft server...")

		go func() {
			err := execDockerCompose(action.Command)
			if err != nil {
				log.Printf("%s", err)
				ctx.Reply(fmt.Sprintf("%s", err))
				return
			}
			time.Sleep(time.Second * 5)
			sess.ChannelMessageEdit(msg.ChannelID, mcMsg.ID, action.Message)
		}()
		return nil
	}
```

File: bot/mc.go [hunk 9]
```diff
@@ -4,43 +4,116 @@ import (
...
+	return nil
 }
 
 func minecraftCommand(command string) (string, error) {
-
 	rconHost := os.Getenv("MC_HOST")
 	rconPort := os.Getenv("RCON_PORT")
 	rconPass := os.Getenv("RCON_PASS")
 
+	if rconHost == "" || rconPort == "" || rconPass == "" {
```
Comment (b1naryth1ef) [inline bot/mc.go:92]: Might be nice to do this at startup in the future, either via [an init function](https://go.dev/doc/effective_go#init) or just within our `main.go` initialization we could call `setupMinecraft()` or the likes.

File: bot/mc.go [hunk 10]
```diff
@@ -4,43 +4,116 @@ import (
...
-	log.Printf("Response %s", response)
-	return response, nil
+	log.Printf("Response %s", resp)
+	return resp, nil
+}
+
+func whitelistCmd(ctx CommandCtx) error {
+	msg := ctx.Message.Message
+	args := ctx.Args
+	if memberHasRole(msg.Member, mcRoleId) || msg.Author.ID == morthisId {
```
Comment (b1naryth1ef) [inline bot/mc.go:123]: Same issue as above with the inverted if statement.

**Review by b1naryth1ef — CHANGES_REQUESTED** [review]
Looks mostly good just a few nits, questions, and minor cleanup suggestions.

File: bot/mc.go [hunk 11]
```diff
@@ -10,18 +10,23 @@ import (
...
 	args := ctx.Args
-	actions := map[string]struct {
-		Message string
-		Command string
-	}{
-		"up":      {"The Minecraft Server has been brought up", "up -d"},
-		"down":    {"The Minecraft server has been brought down", "down"},
-		"restart": {"The Minecraft server has been restarted.", "restart"},
-	}
+	actions := minecraftActions
```
Comment (b1naryth1ef) [inline bot/mc.go]: Remove, just use `minecraftActions[args[0]]` below.

File: bot/roles.go [hunk 12]
```diff
@@ -25,7 +28,10 @@ func roleJoinCmd(ctx CommandCtx) error {
 			return fmt.Errorf("error adding user to role: %w", err)
 		} else {
 			log.Printf("Added user with id: %s (%s) to %s role", msg.Author.ID, msg.Author.Username, roleID)
-			ctx.Reply(fmt.Sprintf("You've been added to the group %s.", args[0]))
+			err := ctx.Reply(fmt.Sprintf("You've been added to the group %s.", args[0]))
```
Comment (b1naryth1ef) [inline bot/roles.go]: Shouldn't most of these `if err != nil { log.Printf(...) }` statements just become `if err != nil { return err }`? You can then handle them and log them (with the command name / args / etc if you want) inside the actual CommandRegistry function.

**Review by b1naryth1ef — CHANGES_REQUESTED** [review]

**Comment by braydencw1** [comment]
I believe I have resolved the issues :D

File: bot/bot.go [hunk 13]
```diff
@@ -36,15 +35,21 @@ func HandleMessageEvents(discord *discordgo.Session, msg *discordgo.MessageCreat
 	if msg.Author.ID == discord.State.User.ID {
 		return
 	}
+	if msg.Author.ID != discord.State.User.ID {
```
Comment (b1naryth1ef) [inline bot/bot.go]: This if statement doesn't make sense after the one on 35, does it? Just remove it and keep the print?

File: bot/command_registry.go [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+	}
+	ctx := CommandCtx{
+		Session: s,
+		Message: msg,
+		Args:    args,
+	}
+	// Found mcCmd and executing
+	go func() {
+		if err := command.fn(ctx); err != nil {
+			if err := ctx.Reply("Your command failed, perhaps there isn't enough arguments."); err != nil {
```
Comment (b1naryth1ef) [inline bot/command_registry.go]: This error message seems bad given we actually check the number of arguments above. Lets just print the error to the user?

File: bot/dnd.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		t, err := time.Parse(layout, args[1])
+		if err != nil {
+			return fmt.Errorf("error parsing date: %w", err)
+		}
+		currRoleId := getMemberDNDRole(msg.Member)
+		if currRoleId == "" {
+			log.Printf("Role not found.")
+			if err := ctx.Reply("Your dnd role is not found in the db."); err != nil {
+				return err
+			}
```
Comment (b1naryth1ef) [inline bot/dnd.go]:
```suggestion
			return ctx.Reply("Your dnd role is not found in the db.")
```

File: bot/dnd.go [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	return nil
+}
+
+func whenIsDndCmd(ctx CommandCtx) error {
+	msg := ctx.Message.Message
+	now := time.Now()
+	currRoleId := getMemberDNDRole(msg.Member)
+	if currRoleId == "" {
+		log.Printf("Could not find Dnd Role")
```
Comment (b1naryth1ef) [inline bot/dnd.go]:
```suggestion
		return ctx.Reply("Could not find DND role")
```

File: bot/dnd.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	dateOfPlay, _, err := db.GetLatestPlayDate(currRoleId)
+	if err != nil {
+		if err := ctx.Reply("Could not find play date information. Perhaps wrong server."); err != nil {
+			return err
+		}
```
Comment (b1naryth1ef) [inline bot/dnd.go]: Stop doing these useless if err checks pls & ty.

```suggestion
		return ctx.Reply("Could not find play date information. Perhaps wrong server.")
```

File: bot/dnd.go [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+		return fmt.Errorf("error parsing latest playdate %w", err)
+	}
+	fmtDate := fmt.Sprint(dateOfPlay.Format("01-02-2006"))
+	if dateOfPlay.Before(now) {
+		if err := ctx.Reply(fmt.Sprintf("There is no date currently set. Your last session was: %s", fmtDate)); err != nil {
+			return err
+		}
```
Comment (b1naryth1ef) [inline bot/dnd.go:48]: grrrrr.

```suggestion
		return ctx.Reply(fmt.Sprintf("There is no date currently set. Your last session was: %s", fmtDate))
```

File: bot/dnd.go [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	if err = ctx.Reply(fmtDate); err != nil {
+		return err
+	}
```
Comment (b1naryth1ef) [inline bot/dnd.go:52]: **grrrrr.**
```suggestion
	return ctx.Reply(fmtDate)
```

File: bot/dnd.go [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+			if err := ctx.Reply("Your dnd role is not found in the db."); err != nil {
+				return err
+			}
+		}
+		if err := db.InsertPlayDate(t, currRoleId); err != nil {
+			return fmt.Errorf("error inserting into table %w", err)
+		}
+		if err = ctx.Reply("The Date has been updated."); err != nil {
+			return err
+		}
```
Comment (b1naryth1ef) [inline bot/dnd.go:28]:
```suggestion
		return ctx.Reply("The Date has been updated.")
```

File: bot/mc.go [hunk 21]
```diff
@@ -4,43 +4,137 @@ import (
...
+
+func manageMinecraftCmd(ctx CommandCtx) error {
+	msg := ctx.Message.Message
+	sess := ctx.Session
+	args := ctx.Args
+	action, exists := minecraftActions[args[0]]
+	if !exists {
+		if err := ctx.Reply("Action is not available."); err != nil {
+			return err
+		}
```
Comment (b1naryth1ef) [inline bot/mc.go:33]:
```suggestion
		return ctx.Reply("Action is not available.")
```

File: bot/mc.go [hunk 22]
```diff
@@ -4,43 +4,137 @@ import (
...
+	}
+	if memberHasRole(msg.Member, mcRoleId) || msg.Author.ID == morthisId {
+		mcMsg, _ := sess.ChannelMessageSend(msg.ChannelID, "Attempting to modify the minecraft server...")
+		go func() {
```
Comment (b1naryth1ef) [inline bot/mc.go:37]: We probably don't need this in a goroutine anymore, do we? The command handler spawns this function in a goroutine so we can just do things inline.

Comment (b1naryth1ef) [reply bot/mc.go:37]: Also this will make error handling easier here, we can just return them.

File: bot/mc.go [hunk 23]
```diff
@@ -4,43 +4,137 @@ import (
...
+	return nil
+}
+
+func mcCmd(ctx CommandCtx) error {
+	m := ctx.Message
+	if m.Author.ID == blueId || m.Author.ID == morthisId {
+		res, err := minecraftCommand(strings.Join(ctx.Args, " "))
+		if err != nil {
+			log.Printf("Err: %s", err)
+			return ctx.Reply("Could not send command, Minecraft might be offline.")
```
Comment (b1naryth1ef) [inline bot/mc.go:62]: You did it right here............. :+1:

File: bot/mc.go [hunk 24]
```diff
@@ -4,43 +4,137 @@ import (
...
+func whitelistCmd(ctx CommandCtx) error {
+	msg := ctx.Message.Message
+	args := ctx.Args
+	if memberHasRole(msg.Member, mcRoleId) || msg.Author.ID == morthisId {
+		log.Printf("Whitelisting, %s ", args[0])
+		res, err := minecraftCommand(fmt.Sprintf("whitelist add %s", args[0]))
+		if err != nil {
+			if err := ctx.Reply("Could not send command, Minecraft might be offline."); err != nil {
+				return err
+			}
```
Comment (b1naryth1ef) [inline bot/mc.go:129]:
```suggestion
			return ctx.Reply("Could not send command, Minecraft might be offline.")
```

File: bot/mc.go [hunk 25]
```diff
@@ -4,43 +4,137 @@ import (
...
+		}
+		if err = ctx.Reply(res); err != nil {
+			return err
+		}
+		if err := ctx.Reply("You're not a minecraft admin."); err != nil {
+			return err
+		}
```
Comment (b1naryth1ef) [inline bot/mc.go]: What in gods name is going on here?

File: bot/roles.go [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+	"fmt"
+	"log"
+	"maps"
+	"slices"
+	"strings"
+)
+
+func roleListCmd(ctx CommandCtx) error {
+	rolesString := strings.Join(slices.Collect(maps.Keys(joinableRolesMap)), ", ")
+	if err := ctx.Reply(fmt.Sprintf("Available roles: %s.\n Available commands: !rjoin & !rleave.", rolesString)); err != nil {
```
Comment (b1naryth1ef) [inline bot/roles.go]: ...

File: bot/roles.go [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+	msg := ctx.Message
+	sess := ctx.Session
+	if roleID, exists := joinableRolesMap[args[0]]; exists {
+		err := sess.GuildMemberRoleAdd(msg.GuildID, msg.Author.ID, roleID)
+		if err != nil {
+			log.Printf("error adding role: %s", err)
+			return fmt.Errorf("error adding user to role: %w", err)
+		} else {
+			log.Printf("Added user with id: %s (%s) to %s role", msg.Author.ID, msg.Author.Username, roleID)
+			if err := ctx.Reply(fmt.Sprintf("You've been added to the group %s.", args[0])); err != nil {
```
Comment (b1naryth1ef) [inline bot/roles.go]: ...

File: bot/roles.go [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+func roleLeaveCmd(ctx CommandCtx) error {
+	args := ctx.Args
+	msg := ctx.Message
+	if roleID, exists := joinableRolesMap[args[0]]; exists {
+		err := ctx.Session.GuildMemberRoleRemove(msg.GuildID, msg.Author.ID, roleID)
+		if err != nil {
+			return fmt.Errorf("error removing role: %w", err)
+		} else {
+			log.Printf("Removed user with id: %s (%s) from %s role", msg.Author.ID, msg.Author.Username, roleID)
+			if err := ctx.Reply(fmt.Sprintf("You've been removed from the group %s.", args[0])); err != nil {
```
Comment (b1naryth1ef) [inline bot/roles.go]: ...

File: bot/roles.go [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func roleLeaveCmd(ctx CommandCtx) error {
+	args := ctx.Args
+	msg := ctx.Message
+	if roleID, exists := joinableRolesMap[args[0]]; exists {
+		err := ctx.Session.GuildMemberRoleRemove(msg.GuildID, msg.Author.ID, roleID)
+		if err != nil {
+			return fmt.Errorf("error removing role: %w", err)
+		} else {
```
Comment (b1naryth1ef) [inline bot/roles.go]: dont need the else here, keeps the following code outside of a block/one indent level less.

**Review by b1naryth1ef — CHANGES_REQUESTED** [review]

File: bot/mc.go (hunk 25 above)
Comment (braydencw1) [reply bot/mc.go]: The linter was complaining about not handling errors

**Pull request merged**